torch/csrc/autograd/python_function.cpp

```diff
@@ -382,87 +356,102 @@ static void _transplant_var(VariableImpl& var, const std::shared_ptr<Function>&
+    }
+    auto it = shared_pairs.find(tensor);
+    if (it != shared_pairs.end()) {
+      // It's explicitly marked as shared via mark_shared_storage
+      return it->second->cdata;
+    }
+    return Variable();
+  };
+
+  // Wraps an output Tensor in a Variable or returns the previous wrapper in
+  // the case of in-place modification.
+  auto wrap_output = [&](at::Tensor data, Variable prev, bool is_modified) -> Variable {
+    if (!prev.defined()) {
+      return make_variable(std::move(data));
+    }
+    if (is_modified) {
+      if (prev.is_leaf() && prev.requires_grad()) {
+        throw std::runtime_error("a leaf Variable that requires grad has been used in an in-place operation.");
+      }
+      // If the input was modified, transplant the grad_fn in the graph:
+      // grad_fn <- variable <- self  ==>  grad_fn <- self <- variable
+      prev.get()->grad.reset();
+      prev.get()->hooks.clear();
+      if (auto grad_acc_fn = prev.get()->grad_accumulator.lock()) {
+        auto grad_acc = dynamic_cast<AccumulateGrad*>(grad_acc_fn.get());
+        grad_acc->variable.reset();
+      }
+      return prev;
+    }
+    // An input has been returned, but it wasn't modified. Return it as a view
+    // so that we can attach a new grad_fn to the Variable.
+    return make_variable_view(std::move(prev), std::move(data));
+  };
+
+  t2var_type output2var;
   for (int i = 0; i < num_outputs; i++) {
     PyObject *output = PyTuple_GET_ITEM(raw_output, i);
-    THPVariable *output_var;
-    auto it = t2var.find(output);
-    if (it == t2var.end()) {
-      // A completely new tensor - just wrap it and continue
-      if (is_volatile) {
-        output_var = (THPVariable*)THPVariable_NewVolatile(output);
-      } else {
-        output_var = (THPVariable*)THPVariable_NewWithFunction(output, cdata);
-      }
+
+    THPVariable* output_var;
+    auto it = output2var.find(output);
+    if (it != output2var.end()) {
+      output_var = it->second;
+      Py_INCREF(output_var);
     } else {
-      // If one of the outputs was also an input tensor it's a bit more complicated.
-      THPVariable *input_var = it->second;
-      auto& input_var_ = input_var->cdata;
-      if (input_var_.grad_fn()) {
-        Py_INCREF(input_var);
-        output_var = input_var;
-        // If it's not a leaf we want to move it in the graph so backprop
-        // will be computed correctly, but only if it was modified. Otherwise
-        // it's better to minimize the number of operations that mutate the graph.
-        // grad_fn <- variable <- self  ==>  grad_fn <- self <- variable
-        if (dirty_inputs.count(output) > 0) {
-          _transplant_var(*input_var_.get(), cdata, i, is_volatile);
-        }
+      // Wrap the output in a Variable
+      bool is_modified = dirty_inputs.count(output) > 0;
+      Variable var = wrap_output(
+          torch::createTensor(output),
+          get_shared_base(output),
+          is_modified);
+      if (is_modified) {
```

Comment:
Correctness of these lines depends on the code of `wrap_output` which is a bit far right now. Can you add a comment saying that it's safe to just call `set_history`, because the only case where we return the original variable is when it was an input and modified, so it will always enter the first case?

Reply:
I've moved most of `set_history` into `make_variable` and moved the rebase call up to wrap_output.